Exception when too many files requested to AtomCache

When the same **AtomCache** is used to request for many structures (not at the same time) an Exception is thrown and the following message is given: **Too many open files**. 

It could indicate a resource or memory leak in **AtomCache**, by maintaining some files open (although the structure has already been downloaded). I get those Exceptions when downloading, in parallel 4 at the same time, more than 2K structures. I do not remember the Exception when I calculated the census the first time (6 months ago), although I am not sure if I was using exactly the same code. Any ideas?

This is the **stacktrace** of the Exception:

```java
java.io.FileNotFoundException: /scratch/cache/data/structures/divided/mmCIF/rl/1rlr.cif.gz (Too many open files)
	at java.io.FileInputStream.open(Native Method) ~[?:1.6.0_38]
	at java.io.FileInputStream.<init>(FileInputStream.java:140) ~[?:1.6.0_38]
	at org.biojava.nbio.core.util.FlatFileCache.addToCache(FlatFileCache.java:63) [classes/:?]
	at org.biojava.nbio.core.util.InputStreamProvider.getInputStreamFromFile(InputStreamProvider.java:247) [classes/:?]
	at org.biojava.nbio.core.util.InputStreamProvider.getInputStream(InputStreamProvider.java:163) [classes/:?]
	at org.biojava.nbio.structure.io.LocalPDBDirectory.getInputStream(LocalPDBDirectory.java:377) [classes/:?]
	at org.biojava.nbio.structure.io.LocalPDBDirectory.getStructureById(LocalPDBDirectory.java:343) [classes/:?]
	at org.biojava.nbio.structure.align.util.AtomCache.loadStructureFromCifByPdbId(AtomCache.java:1038) [classes/:?]
	at org.biojava.nbio.structure.align.util.AtomCache.getStructureForPdbId(AtomCache.java:1019) [classes/:?]
	at org.biojava.nbio.structure.scop.ScopDomain.loadStructure(ScopDomain.java:245) [classes/:?]
	at org.biojava.nbio.structure.align.client.StructureName.loadStructure(StructureName.java:349) [classes/:?]
	at org.biojava.nbio.structure.align.util.AtomCache.getStructure(AtomCache.java:394) [classes/:?]
	at demo.CeSymmMain$CeSymmWorker.run(CeSymmMain.java:1061) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [?:1.6.0_38]
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334) [?:1.6.0_38]
	at java.util.concurrent.FutureTask.run(FutureTask.java:166) [?:1.6.0_38]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146) [?:1.6.0_38]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [?:1.6.0_38]
	at java.lang.Thread.run(Thread.java:701) [?:1.6.0_38]
``

This Exception is also thrown some of the times, in the same case:

```java
Caused by: java.net.SocketException: Too many open files
	at java.net.Socket.createImpl(Socket.java:414) ~[?:1.6.0_38]
	at java.net.Socket.connect(Socket.java:544) ~[?:1.6.0_38]
	at sun.net.NetworkClient.doConnect(NetworkClient.java:173) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:427) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:529) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:213) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.New(HttpClient.java:306) ~[?:1.6.0_38]
	at sun.net.www.http.HttpClient.New(HttpClient.java:325) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:956) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:892) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:810) ~[?:1.6.0_38]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1227) ~[?:1.6.0_38]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:397) ~[?:1.6.0_38]
	at org.biojava.nbio.structure.align.util.HTTPConnectionTools.getInputStream(HTTPConnectionTools.java:186) ~[classes/:?]
	at org.biojava.nbio.structure.align.util.HTTPConnectionTools.getInputStream(HTTPConnectionTools.java:158) ~[classes/:?]
	at org.biojava.nbio.structure.scop.RemoteScopInstallation.requestRemoteDomainByScopID(RemoteScopInstallation.java:208) ~[classes/:?]
	at org.biojava.nbio.structure.scop.RemoteScopInstallation.getDomainByScopID(RemoteScopInstallation.java:220) ~[classes/:?]
```

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ChemCompGroupFactory.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ChemCompGroupFactory.java
@@ -57,10 +57,8 @@ public class ChemCompGroupFactory {
 		logger.debug("Chem comp "+recordName+" read from provider "+chemCompProvider.getClass().getCanonicalName());
 		cc = chemCompProvider.getChemComp(recordName);
 		
-		// If chemCompProvider fails don't try to cache null & one letter code may be null.
-		if (null != cc && !cc.isEmpty()){
-			cache.put(recordName, cc);
-		}
+		// Note that this also caches null or empty responses
+		cache.put(recordName, cc);
 		return cc;
 	}
 
@@ -79,6 +77,8 @@ public class ChemCompGroupFactory {
 	public static void setChemCompProvider(ChemCompProvider provider) {
 		logger.debug("Setting new chem comp provider to "+provider.getClass().getCanonicalName());
 		chemCompProvider = provider;
+		// clear cache
+		cache.clear();
 	}
 
 	public static ChemCompProvider getChemCompProvider(){
@@ -153,10 +153,5 @@ public class ChemCompGroupFactory {
 		}
 		return oneLetter;
 	}
-
-	public static SoftHashMap<String, ChemComp> getCache() {
-		return cache;
-	}
-
 
 }

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ChemCompGroupFactory.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ChemCompGroupFactory.java
@@ -58,7 +58,7 @@ public class ChemCompGroupFactory {
 		cc = chemCompProvider.getChemComp(recordName);
 		
 		// If chemCompProvider fails don't try to cache null & one letter code may be null.
-		if (null != cc && !"?".equals(cc.getOne_letter_code())){
+		if (null != cc && !cc.isEmpty()){
 			cache.put(recordName, cc);
 		}
 		return cc;

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/DownloadChemCompProvider.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/DownloadChemCompProvider.java
@@ -166,62 +166,57 @@ public class DownloadChemCompProvider implements ChemCompProvider {
 		int counter = 0;
 		InputStreamProvider prov = new InputStreamProvider();
 
-		InputStream inStream = prov.getInputStream(f);
+		try( BufferedReader buf = new BufferedReader (new InputStreamReader (prov.getInputStream(f)));
+				) {
+			String line = null;
+			line = buf.readLine ();
+			StringWriter writer = new StringWriter();
 
-		BufferedReader buf = new BufferedReader (new InputStreamReader (inStream));
+			String currentID = null;
+			while (line != null){
 
-		String line = null;
-		line = buf.readLine ();
-		StringWriter writer = new StringWriter();
+				if ( line.startsWith("data_")) {
+					// a new record found!
 
-		String currentID = null;
-		while (line != null){
+					if ( currentID != null) {
+						writeID(writer.toString(), currentID);
+						counter++;
+					}
 
-			if ( line.startsWith("data_")) {
-				// a new record found!
-
-				if ( currentID != null) {
-					writeID(writer, currentID);
-					counter++;
+					currentID = line.substring(5);
+					writer = new StringWriter();
 				}
 
-				currentID = line.substring(5);
-				writer = new StringWriter();
+				writer.append(line);
+				writer.append(NEWLINE);
+
+				line = buf.readLine ();
 			}
 
-			writer.append(line);
-			writer.append(NEWLINE);
+			// write the last record...
+			writeID(writer.toString(),currentID);
+			counter++;
 
-			line = buf.readLine ();
 		}
-
-		// write the last record...
-		writeID(writer,currentID);
-		counter++;
-		
-		inStream.close();
 
 		logger.info("Created " + counter + " chemical component files.");
 	}
 
-	private void writeID(StringWriter writer, String currentID) throws IOException{
+	/**
+	 * Output chemical contents to a file
+	 * @param contents File contents
+	 * @param currentID Chemical ID, used to determine the filename
+	 * @throws IOException
+	 */
+	private void writeID(String contents, String currentID) throws IOException{
 
-		
 		String localName = DownloadChemCompProvider.getLocalFileName(currentID);
 
-		FileOutputStream outPut = new FileOutputStream(localName);
+		try ( PrintWriter pw = new PrintWriter(new GZIPOutputStream(new FileOutputStream(localName))) ) {
 
-		GZIPOutputStream gzOutPut = new GZIPOutputStream(outPut);
-
-		PrintWriter pw = new PrintWriter(gzOutPut);
-
-		pw.print(writer.toString());
-		writer.close();
-		pw.flush();
-		pw.close();
-
-		outPut.close();
-
+			pw.print(contents.toString());
+			pw.flush();
+		}
 	}
 
 	/** 
@@ -354,7 +349,7 @@ public class DownloadChemCompProvider implements ChemCompProvider {
 
 		String u = SERVER_LOCATION + recordName + ".cif";
 
-//		System.out.println("downloading " + u);
+		logger.debug("downloading " + u);
 
 		URL url = null;
 		
@@ -364,29 +359,20 @@ public class DownloadChemCompProvider implements ChemCompProvider {
 
 			HttpURLConnection uconn = HTTPConnectionTools.openHttpURLConnection(url);
 
-			InputStream conn = uconn.getInputStream();
+			try( PrintWriter pw = new PrintWriter(new GZIPOutputStream(new FileOutputStream(localName)));
+					BufferedReader fileBuffer = new BufferedReader(new InputStreamReader(uconn.getInputStream()));
+					) {
 
-			FileOutputStream outPut = new FileOutputStream(localName);
+				String line;
 
-			GZIPOutputStream gzOutPut = new GZIPOutputStream(outPut);
+				while ((line = fileBuffer.readLine()) != null) {
+					pw.println(line);
+				}
 
-			PrintWriter pw = new PrintWriter(gzOutPut);
+				pw.flush();
 
-			BufferedReader fileBuffer = new BufferedReader(new InputStreamReader(conn));
-
-			String line;
-
-			while ((line = fileBuffer.readLine()) != null) {
-				pw.println(line);
+				return true;
 			}
-
-			pw.flush();
-			pw.close();
-
-			outPut.close();
-			conn.close();
-
-			return true;
 		} catch (FileNotFoundException e) {
 			// Possible that there is no ChemComp matching this group.
 			logger.warn(recordName + " is not available from " + SERVER_LOCATION + " and will be skipped");

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ReducedChemCompProvider.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ReducedChemCompProvider.java
@@ -20,17 +20,15 @@
  */
 package org.biojava.nbio.structure.io.mmcif;
 
-import org.biojava.nbio.structure.io.mmcif.chem.PolymerType;
-import org.biojava.nbio.structure.io.mmcif.chem.ResidueType;
-import org.biojava.nbio.structure.io.mmcif.model.ChemComp;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.util.zip.GZIPInputStream;
+
+import org.biojava.nbio.structure.io.mmcif.model.ChemComp;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /** Unlike the {@link DownloadChemCompProvider}, this  {@link ChemCompProvider} does not download any chem comp definitions. 
@@ -51,36 +49,35 @@ public class ReducedChemCompProvider implements ChemCompProvider {
 	@Override
 	public ChemComp getChemComp(String recordName) {
 		String name = recordName.toUpperCase().trim();
-		try {
-			try(InputStream inStream = this.getClass().getResourceAsStream("/chemcomp/"+name + ".cif.gz")) {
+		try(InputStream inStream = this.getClass().getResourceAsStream("/chemcomp/"+name + ".cif.gz")) {
 
-				logger.debug("Reading chemcomp/"+name+".cif.gz");
-				
-				if ( inStream == null){
-					//System.out.println("Could not find chem comp: " + name + " ... using generic Chem Comp");
-					// could not find the chem comp definition for this in the jar file
-					logger.debug("Getting empty chem comp for {}",name);
-					ChemComp cc = ChemComp.getEmptyChemComp();
-					cc.setId(name);
-					return cc;
-				}
+			logger.debug("Reading chemcomp/"+name+".cif.gz");
 
-				MMcifParser parser = new SimpleMMcifParser();
-
-				ChemCompConsumer consumer = new ChemCompConsumer();
-
-				// The Consumer builds up the BioJava - structure object.
-				// you could also hook in your own and build up you own data model.
-				parser.addMMcifConsumer(consumer);
-
-				parser.parse(new BufferedReader(new InputStreamReader(new GZIPInputStream(inStream))));
-
-				ChemicalComponentDictionary dict = consumer.getDictionary();
-
-				ChemComp chemComp = dict.getChemComp(name);
-
-				return chemComp;
+			if ( inStream == null){
+				//System.out.println("Could not find chem comp: " + name + " ... using generic Chem Comp");
+				// could not find the chem comp definition for this in the jar file
+				logger.debug("Getting empty chem comp for {}",name);
+				ChemComp cc = ChemComp.getEmptyChemComp();
+				cc.setId(name);
+				return cc;
 			}
+
+			MMcifParser parser = new SimpleMMcifParser();
+
+			ChemCompConsumer consumer = new ChemCompConsumer();
+
+			// The Consumer builds up the BioJava - structure object.
+			// you could also hook in your own and build up you own data model.
+			parser.addMMcifConsumer(consumer);
+
+			parser.parse(new BufferedReader(new InputStreamReader(new GZIPInputStream(inStream))));
+
+			ChemicalComponentDictionary dict = consumer.getDictionary();
+
+			ChemComp chemComp = dict.getChemComp(name);
+
+			return chemComp;
+
 		} catch (IOException e){
 			logger.error("IOException caught while reading chem comp {}.",name,e);
 		}

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ReducedChemCompProvider.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/ReducedChemCompProvider.java
@@ -46,54 +46,46 @@ public class ReducedChemCompProvider implements ChemCompProvider {
 	public ReducedChemCompProvider(){
 		logger.debug("Initialising ReducedChemCompProvider");
 	}
-	
-	
-	public ChemComp getEmptyChemComp(){
-		ChemComp comp = new ChemComp();
-		
-		comp.setOne_letter_code("?");
-		comp.setPolymerType(PolymerType.unknown);
-		comp.setResidueType(ResidueType.atomn);
-		return comp;
-	}
-	
+
+
 	@Override
 	public ChemComp getChemComp(String recordName) {
 		String name = recordName.toUpperCase().trim();
-		InputStream inStream = this.getClass().getResourceAsStream("/chemcomp/"+name + ".cif.gz");
-
-		if ( inStream == null){
-			//System.out.println("Could not find chem comp: " + name + " ... using generic Chem Comp");
-			// could not find the chem comp definition for this in the jar file
-			logger.debug("Getting empty chem comp for {}",name);
-			ChemComp cc = getEmptyChemComp();
-			cc.setId(name);
-			return cc;
-		}
-
-		MMcifParser parser = new SimpleMMcifParser();
-
-		ChemCompConsumer consumer = new ChemCompConsumer();
-
-		// The Consumer builds up the BioJava - structure object.
-		// you could also hook in your own and build up you own data model.
-		parser.addMMcifConsumer(consumer);
-
 		try {
-			parser.parse(new BufferedReader(new InputStreamReader(new GZIPInputStream(inStream))));
+			try(InputStream inStream = this.getClass().getResourceAsStream("/chemcomp/"+name + ".cif.gz")) {
 
-			ChemicalComponentDictionary dict = consumer.getDictionary();
+				logger.debug("Reading chemcomp/"+name+".cif.gz");
+				
+				if ( inStream == null){
+					//System.out.println("Could not find chem comp: " + name + " ... using generic Chem Comp");
+					// could not find the chem comp definition for this in the jar file
+					logger.debug("Getting empty chem comp for {}",name);
+					ChemComp cc = ChemComp.getEmptyChemComp();
+					cc.setId(name);
+					return cc;
+				}
 
-			ChemComp chemComp = dict.getChemComp(name);
-			
-			return chemComp;
+				MMcifParser parser = new SimpleMMcifParser();
+
+				ChemCompConsumer consumer = new ChemCompConsumer();
+
+				// The Consumer builds up the BioJava - structure object.
+				// you could also hook in your own and build up you own data model.
+				parser.addMMcifConsumer(consumer);
+
+				parser.parse(new BufferedReader(new InputStreamReader(new GZIPInputStream(inStream))));
+
+				ChemicalComponentDictionary dict = consumer.getDictionary();
+
+				ChemComp chemComp = dict.getChemComp(name);
+
+				return chemComp;
+			}
 		} catch (IOException e){
-			logger.error("IOException caught while reading chem comp {}. Error: {}",name,e.getMessage());
-			//e.printStackTrace();
-
+			logger.error("IOException caught while reading chem comp {}.",name,e);
 		}
 		logger.warn("Problem when loading chem comp {}, will use an empty chem comp for it", name);
-		ChemComp cc = getEmptyChemComp();
+		ChemComp cc = ChemComp.getEmptyChemComp();
 		cc.setId(name);
 		return cc;
 	}

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/model/ChemComp.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/model/ChemComp.java
@@ -594,7 +594,7 @@ public class ChemComp implements Serializable, Comparable<ChemComp>{
 		ChemComp comp = new ChemComp();
 
 		comp.setOne_letter_code("?");
-		comp.setThree_letter_code("???");
+		comp.setThree_letter_code("???"); // Main signal for isEmpty()
 		comp.setPolymerType(PolymerType.unknown);
 		comp.setResidueType(ResidueType.atomn);
 		return comp;
@@ -606,7 +606,7 @@ public class ChemComp implements Serializable, Comparable<ChemComp>{
 	 */
 	public boolean isEmpty() {
 		// Is this the best flag for it being empty?
-		return id == null || getThree_letter_code().equals("???");
+		return id == null || getThree_letter_code() == null || getThree_letter_code().equals("???");
 	}
 
 }

--- a/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/model/ChemComp.java
+++ b/biojava-structure/src/main/java/org/biojava/nbio/structure/io/mmcif/model/ChemComp.java
@@ -40,40 +40,40 @@ public class ChemComp implements Serializable, Comparable<ChemComp>{
 	 */
 	private static final long serialVersionUID = -4736341142030215915L;
 
-	String id ;
-	String name;
-	String type;
-	String pdbx_type;
-	String formula;
-	String mon_nstd_parent_comp_id;
-	String pdbx_synonyms;
-	String pdbx_formal_charge;
-	String pdbx_initial_date ;
-	String pdbx_modified_date;
-	String pdbx_ambiguous_flag;
-	String pdbx_release_status ;
-	String pdbx_replaced_by;
-	String pdbx_replaces;
-	String formula_weight;
-	String one_letter_code;
-	String three_letter_code;
-	String pdbx_model_coordinates_details;
-	String pdbx_model_coordinates_missing_flag;
-	String pdbx_ideal_coordinates_details;
-	String pdbx_ideal_coordinates_missing_flag;
-	String pdbx_model_coordinates_db_code;
-	String pdbx_subcomponent_list;
-	String pdbx_processing_site;
-	String mon_nstd_flag;
+	private String id ;
+	private String name;
+	private String type;
+	private String pdbx_type;
+	private String formula;
+	private String mon_nstd_parent_comp_id;
+	private String pdbx_synonyms;
+	private String pdbx_formal_charge;
+	private String pdbx_initial_date ;
+	private String pdbx_modified_date;
+	private String pdbx_ambiguous_flag;
+	private String pdbx_release_status ;
+	private String pdbx_replaced_by;
+	private String pdbx_replaces;
+	private String formula_weight;
+	private String one_letter_code;
+	private String three_letter_code;
+	private String pdbx_model_coordinates_details;
+	private String pdbx_model_coordinates_missing_flag;
+	private String pdbx_ideal_coordinates_details;
+	private String pdbx_ideal_coordinates_missing_flag;
+	private String pdbx_model_coordinates_db_code;
+	private String pdbx_subcomponent_list;
+	private String pdbx_processing_site;
+	private String mon_nstd_flag;
 
-	List<ChemCompDescriptor> descriptors = new ArrayList<ChemCompDescriptor>();
-	List<ChemCompBond> bonds = new ArrayList<ChemCompBond>();
-	List<ChemCompAtom> atoms = new ArrayList<ChemCompAtom>();
+	private List<ChemCompDescriptor> descriptors = new ArrayList<ChemCompDescriptor>();
+	private List<ChemCompBond> bonds = new ArrayList<ChemCompBond>();
+	private List<ChemCompAtom> atoms = new ArrayList<ChemCompAtom>();
 
 	// and some derived data for easier processing...
-	ResidueType residueType;
-	PolymerType polymerType;
-	boolean standard;
+	private ResidueType residueType;
+	private PolymerType polymerType;
+	private boolean standard;
 
 	@Override
 	public String toString(){
@@ -585,6 +585,28 @@ public class ChemComp implements Serializable, Comparable<ChemComp>{
 			return false;
 		return true;
 	}
+	
+	/**
+	 * Creates a new instance of the dummy empty ChemComp.
+	 * @return
+	 */
+	public static ChemComp getEmptyChemComp(){
+		ChemComp comp = new ChemComp();
 
+		comp.setOne_letter_code("?");
+		comp.setThree_letter_code("???");
+		comp.setPolymerType(PolymerType.unknown);
+		comp.setResidueType(ResidueType.atomn);
+		return comp;
+	}
+
+	/**
+	 * Indicates whether this compound was created with 
+	 * @return
+	 */
+	public boolean isEmpty() {
+		// Is this the best flag for it being empty?
+		return id == null || getThree_letter_code().equals("???");
+	}
 
 }

--- a/biojava-structure/src/test/java/org/biojava/nbio/structure/ChemCompTest.java
+++ b/biojava-structure/src/test/java/org/biojava/nbio/structure/ChemCompTest.java
@@ -117,10 +117,6 @@ public class ChemCompTest {
 	@Test
     public void testChangingProviders(){
 
-		//reset the cache
-		SoftHashMap<String, ChemComp> cache = ChemCompGroupFactory.getCache();
-		cache.clear();
-		
 		// test for issue #145
 
         String chemID = "MEA";
@@ -134,7 +130,7 @@ public class ChemCompTest {
 
         assertTrue(" is not mea" , cc.getId().equals(chemID));
         
-        assertNull(cc.getThree_letter_code());
+        assertTrue(cc.isEmpty());
 
         // now we change to download chem comp provider
 
@@ -151,7 +147,6 @@ public class ChemCompTest {
         
         
         // now testing in opposite order
-        cache.clear();
 
         // first we test with download chem comp provider
 

